### PR TITLE
Add attn to discriminator.

### PR DESCRIPTION
**Pull request by mrdrozdov**
This branch implements an example of visual attention within the Discriminator of DCGAN. This approach is similar to [1] except that attention is done with a convolutional layer rather than a linear layer.

[1] Xu et al. Show, Attend and Tell: Neural Image Caption Generation with Visual Attention. https://arxiv.org/pdf/1502.03044.pdf (code: https://github.com/kelvinxu/arctic-captions/blob/master/capgen.py)

Interesting features to add (not included):

- [ ] Use entropy of the attention mask as a regularization term. The intuition is that most of the image should not make the discriminator's job easy. Entropy is \sum p(x) log p(x), and p(x) here would be the attention value for each pixel.
- [ ] Keep the attention mask from the previous N batches. Give this to the generator so that it can determine which part of the image it needs to work on the most. There might be some work recently that is similar to this where two networks communicate their gradients.

### CODE REVIEW

File: dcgan/main.py [hunk 1]
```diff
@@ -148,32 +148,50 @@ class _netD(nn.Module):
...
-            nn.Conv2d(ndf * 4, ndf * 8, 4, 2, 1, bias=False),
-            nn.BatchNorm2d(ndf * 8),
-            nn.LeakyReLU(0.2, inplace=True),
-            # state size. (ndf*8) x 4 x 4
-            nn.Conv2d(ndf * 8, 1, 4, 1, 0, bias=False),
-            nn.Sigmoid()
-        )
+        # attention
+        self.conv_attn = nn.Conv2d(nc, 1, 3, 1, 1, bias=False)
+        self.sigmoid_attn = nn.Sigmoid()
```
Comment (mrdrozdov) [inline dcgan/main.py:153]: I add the attention layer as a combination of a convolutional layer and sigmoid.

File: dcgan/main.py [hunk 2]
```diff
@@ -148,32 +148,50 @@ class _netD(nn.Module):
...
     def forward(self, input):
-        if isinstance(input.data, torch.cuda.FloatTensor) and self.ngpu > 1:
-            output = nn.parallel.data_parallel(self.main, input, range(self.ngpu))
-        else:
-            output = self.main(input)
+        x = input
+        # print(x.size())  # (64L, 3L, 64L, 64L)
+        attn = self.conv_attn(x)
+        attn = self.sigmoid_attn(attn)
+        x = x * attn
```
Comment (mrdrozdov) [inline dcgan/main.py:178]: This is where attention is applied.

File: dcgan/main.py [hunk 3]
```diff
@@ -148,32 +148,50 @@ class _netD(nn.Module):
...
+        x = self.bn2(x)
+        x = self.relu2(x)
+        x = self.conv3(x)
+        x = self.bn3(x)
+        x = self.relu3(x)
+        x = self.conv4(x)
+        x = self.bn4(x)
+        x = self.relu4(x)
+        x = self.conv5(x)
+        x = self.sigmoid5(x)
```
Comment (mrdrozdov) [inline dcgan/main.py:193]: I removed the use of `nn.Sequential`. Instead, we call each layer explicitly. This is more verbose but makes debugging easier (specifically, setting break points and printing the size of data).

**Comment by mrdrozdov** [comment]
Might not be deep enough!
